helix-view/src/register.rs

```diff
@@ -0,0 +1,369 @@
+        self.values = values;
+        save_to_clipboard(&self.values, editor, ClipboardType::Clipboard)
+    }
+
+    fn push(&mut self, editor: &mut Editor, value: String) -> Result<()> {
+        self.values.push(value);
+        save_to_clipboard(&self.values, editor, ClipboardType::Clipboard)
+    }
+}
+
+#[derive(Debug, Default)]
+struct PrimaryClipboardRegister {
+    values: Vec<String>,
+}
+
+impl Register for PrimaryClipboardRegister {
+    fn name(&self) -> char {
+        '+'
+    }
+
+    fn preview(&self) -> &str {
+        "<primary clipboard>"
+    }
+
+    fn read(&self, editor: &Editor) -> Vec<String> {
+        read_from_clipboard(&self.values, editor, ClipboardType::Selection)
+    }
+
+    fn write(&mut self, editor: &mut Editor, values: Vec<String>) -> Result<()> {
+        self.values = values;
+        save_to_clipboard(&self.values, editor, ClipboardType::Selection)
+    }
+
+    fn push(&mut self, editor: &mut Editor, value: String) -> Result<()> {
+        self.values.push(value);
+        save_to_clipboard(&self.values, editor, ClipboardType::Selection)
+    }
+}
+
+fn save_to_clipboard(
+    values: &[String],
+    editor: &mut Editor,
+    clipboard_type: ClipboardType,
+) -> Result<()> {
+    let line_ending = doc!(editor).line_ending;
+    let joined = values.join(line_ending.as_str());
+
+    editor
+        .clipboard_provider
+        .set_contents(joined, clipboard_type)
+}
+
+fn read_from_clipboard(
+    saved_values: &[String],
+    editor: &Editor,
+    clipboard_type: ClipboardType,
+) -> Vec<String> {
+    match editor.clipboard_provider.get_contents(clipboard_type) {
+        Ok(contents) => {
+            // If we're pasting the same value that we just yanked, re-use
+            // the saved values. This allows pasting multiple selections
+            // even when yanked to a clipboard.
+            if contents_are_saved(saved_values, editor, &contents) {
+                saved_values.to_owned()
+            } else {
+                vec![contents]
+            }
+        }
+        Err(err) => {
+            log::error!(
+                "Failed to read {} clipboard: {}",
+                match clipboard_type {
+                    ClipboardType::Clipboard => "system",
+                    ClipboardType::Selection => "primary",
+                },
+                err
+            );
+
+            Vec::new()
+        }
+    }
+}
+
+fn contents_are_saved(saved_values: &[String], editor: &Editor, mut contents: &str) -> bool {
+    let line_ending = doc!(editor).line_ending.as_str();
+    let mut values = saved_values.iter();
+
+    match values.next() {
+        Some(first) if contents.starts_with(first) => {
+            contents = &contents[first.len()..];
+        }
+        _ => return false,
+    }
+
+    for value in values {
+        if contents.starts_with(line_ending) == contents[line_ending.len()..].starts_with(value) {
```

Comment:
Shouldn't this be:

```suggestion
        if contents.starts_with(line_ending) && contents[line_ending.len()..].starts_with(value) {
```

If document line ending "\r" and content starts with "\n", hen the if statement is true when the remaining content isn't the same right? (false == false => true)

Reply:
Might be wrong, but the entire function body might be replaced with something similar to this:
 ```rust
cached_value.join(editor.focused_document().line_ending.as_str()) == value
```

Reply:
`join` alloates a new string so that's pretty inefficient which is why we iterate instead, You are correct about the and tough that looks like a typo